adapters/advangelists/advangelists.go

```diff
@@ -0,0 +1,252 @@
+	if len(request.Imp) == 0 {
+		errs = append(errs, &errortypes.BadInput{Message: "No impression in the bid request"})
+		return nil, errs
+	}
+	imps, impExts, err := getImpressionsInfo(request.Imp)
+	if len(imps) == 0 {
+		return nil, err
+	}
+	errs = append(errs, err...)
+
+	pub2impressions, dispErrors := dispatchImpressions(imps, impExts)
+	if len(pub2impressions) == 0 {
+		return nil, errs
+	}
+	errs = append(errs, dispErrors...)
+
+	result := make([]*adapters.RequestData, 0, len(pub2impressions))
+	for k, imps := range pub2impressions {
+		bidRequest, err := adapter.buildAdapterRequest(request, &k, imps)
+		if err != nil {
+			errs = append(errs, err)
+		} else {
+			result = append(result, bidRequest)
+		}
+	}
+	return result, errs
+}
+
+// getImpressionsInfo checks each impression for validity and returns impressions copy with corresponding exts
+func getImpressionsInfo(imps []openrtb.Imp) ([]openrtb.Imp, []openrtb_ext.ExtImpAdvangelists, []error) {
+	impsCount := len(imps)
+	errors := make([]error, 0, impsCount)
+	resImps := make([]openrtb.Imp, 0, impsCount)
+	resImpExts := make([]openrtb_ext.ExtImpAdvangelists, 0, impsCount)
+
+	for _, imp := range imps {
+		impExt, err := getImpressionExt(&imp)
+		if err != nil {
+			errors = append(errors, err)
+			continue
+		}
+		if err := validateImpression(&imp, impExt); err != nil {
+			errors = append(errors, err)
+			continue
+		}
+		resImps = append(resImps, imp)
+		resImpExts = append(resImpExts, *impExt)
+
+	}
+	return resImps, resImpExts, errors
+}
+
+func validateImpression(imp *openrtb.Imp, impExt *openrtb_ext.ExtImpAdvangelists) error {
+	if impExt.PubID == "" {
+		return &errortypes.BadInput{Message: "No pubid value provided"}
+	}
+	return nil
+}
+
+//Group impressions by advangelists-specific parameters `pubid
+func dispatchImpressions(imps []openrtb.Imp, impsExt []openrtb_ext.ExtImpAdvangelists) (map[openrtb_ext.ExtImpAdvangelists][]openrtb.Imp, []error) {
+	res := make(map[openrtb_ext.ExtImpAdvangelists][]openrtb.Imp)
+	errors := make([]error, 0)
+	for idx := range imps {
+		imp := imps[idx]
+		err := compatImpression(&imp)
+		if err != nil {
+			errors = append(errors, err)
+			continue
+		}
+		impExt := impsExt[idx]
+		if res[impExt] == nil {
+			res[impExt] = make([]openrtb.Imp, 0)
+		}
+		res[impExt] = append(res[impExt], imp)
+
+	}
+	return res, errors
+}
+
+//Alter impression info to comply with advangelists platform requirements
+func compatImpression(imp *openrtb.Imp) error {
+	imp.Ext = nil //do not forward ext to advangelists platform
+	if imp.Banner != nil {
+		return compatBannerImpression(imp)
+	}
+	return nil
+}
+
+func compatBannerImpression(imp *openrtb.Imp) error {
+	// Create a copy of the banner, since imp is a shallow copy of the original.
+	bannerCopy := *imp.Banner
+	banner := &bannerCopy
+	//As banner.w/h are required fields for advangelistsAdn platform - take the first format entry
+	if banner.W == nil && banner.H == nil {
```

Comment:
Are BOTH fields required? In other words, if the truth table below is accurate, shouldn't we switch the boolean operator to a short circuit OR ( || )?
```
banner.W |  banner.H |   if statement
-----------------------------------------
  nil    |    nil    | into true branch 
  nil    |  not nil  | into true branch 
 not nil |    nil    | into true branch 
 not nil |  not nil  |    skip          
```

Reply:
agree, updated. thanks